### ISSUE TITLE
how to trace a application which is run by using a tcserver which runs tomcat inbuilt

### ISSUE THREAD

**Issue by AbhishekPuligundla** [jbachorik/btrace#169]
i want to trace a application which is run by using a tcserver. But it is not allowing the btrace to connect with the port. Normal tomcat applications are being traced but the application which use tcserver is not possible to trace. #168 is the type of error i am getting when i try to trace the application with tcserver.
I tried to print the debug information while running the applicaiton and i am getting the following result
DEBUG: accepting port 2024
DEBUG: assuming default classpath '.'
.
.
.
.
DEBUG: loaded /usr/java/btracebin/build/btrace-agent.jar
DEBUG: registering shutdown hook
DEBUG: registering signal handler for SIGINT
DEBUG: submitting the BTrace program
DEBUG: opening socket to 2024
DEBUG: server not yet available; retrying ...
DEBUG: server not yet available; retrying ...
DEBUG: server not yet available; retrying ...
DEBUG: server not yet available; retrying ...
DEBUG: server not yet available; retrying ...
DEBUG: server not yet available; retrying ...
DEBUG: server not yet available; retrying ...
DEBUG: server not yet available; retrying ...
.
.
.
.
.
.
.
DEBUG: server not yet available; retrying ...
DEBUG: server not yet available; retrying ...
DEBUG: server not yet available; retrying ...
DEBUG: server not yet available; retrying ...
Exception in thread "main" java.lang.NullPointerException
        at com.sun.btrace.client.Client.submit(Client.java:353)
        at com.sun.btrace.client.Main.main(Main.java:189)
DEBUG: sending exit command
Exception in thread "Thread-0" java.lang.IllegalStateException
        at com.sun.btrace.client.Client.send(Client.java:458)
        at com.sun.btrace.client.Client.sendExit(Client.java:392)
        at com.sun.btrace.client.Main$2.run(Main.java:226)
        at java.lang.Thread.run(Thread.java:745)

**Comment by jbachorik** [comment]
Once you see the 'DEBUG: accepting port 2024' message can you telnet to
port 2024 on the machine the tcserver is running on?
This looks like a firewall blocking communication - BTrace client simply
can not create connection to port 2024.

**Comment by jbachorik** [comment]
Any luck with telneting to BTrace port?

**Comment by AbhishekPuligundla** [comment]
I find that the inbuilt tomcat application itself as a jre which is different from the system java and so when trying to attach btrace agent , the btrace is looking for tools.jar file in the jre of the application rather than using the system. so i just added the tools.jar file in the respective path and it's working. i would like to ask why btrace is not using the tools.jar of the system and looking for the tools.jar file in the application jre.

**Comment by jbachorik** [comment]
Well, this is simple. 'tools.jar' is implementation dependent so using a
version which does not come with the JVM you are trying to attach to might
cause problems.

**Comment by bvanders** [comment]
@AbhishekPuligundla are you implying that you added the tools.jar to the tomcat lib directory?

**Comment by AbhishekPuligundla** [comment]
not to the tomcat cat lib directory but to the jre folder of the application

**Comment by bvanders** [comment]
@AbhishekPuligundla that worked in my testing.  Thanks.  I would prefer if there was a way to instruct btrace to use the tools.jar in the non-JRE lib directory.  That is, I have the jar present in my $JAVA_HOME/lib, but it is not present by default in $JAVA_HOME/jre/lib.  @jbachorik can I point btrace at the "right" directory?

**Comment by jbachorik** [comment]
@bvanders No, it's not currently possible. Please, open an enhancement issue for this - shouldn't be that difficult to implement.

**Comment by moonlitSpider** [comment]
I wonder if I might ask @jbachorik to re-open this issue because I am hitting it reliably.

I lately downloaded & installed btrace 1.2.5.1 into an Ubuntu VM (VMware). What's perhaps problematic about my environment is that the JVM I am trying to run btrace against is running in a Docker container. The JVM itself is running a Virgo OSGI server.

btrace runs on the host (not in a container). But I've done two things that should enable btrace to access the container.

1. run the container with Docker's pid=host flag. This allows the container to share the host's PID space (no pid isolation). I know that this was effective because, before enabling this, I couldn't run JVisualVM against the JVM. But now I can.

2. Expose container port 2024 as host port 2024. I am not at all sure what this does, or why it's even necessary.

In any event, I am getting error "Unable to open socket file: target process not responding or HotSpot VM not loaded" and would be most grateful for some debugging guidance on this. I'd very much like to get btrace working.

Java is JDK1.7.0_80.

Thank you.

-Paul

**Comment by jbachorik** [comment]
@moonlitSpider The mention of OSGI server is new here - that may very well be causing the problems. You need to configure the OSGI container to play nicely with BTrace - see [this gist](https://gist.github.com/jbachorik/82f8c2a0d9574af14d19b45704773187#file-btrace_osgi-md)

The BTrace version you are using is rather old - the current version is 1.3.6 so you might try that one.

Regarding the port - BTrace socket server is by default running on port 2020. I assume the 2024 value must be a remnant of someones customized setup?

**Comment by moonlitSpider** [comment]
Dear Mr. Bachorik,


Thank you for your reply. Most kind of you.


Where can I download the 1.3.6 version?


Cordially,


Paul

________________________________

**Comment by moonlitSpider** [comment]
Ah, never mind; thanks. I found it at https://github.com/btraceio/btrace/releases/tag/v1.3.6

I will try again with this version.

Cordially,

Paul

**Comment by jbachorik** [comment]
You can always use https://github.com/btraceio/btrace/releases/latest to
get the latest released version.

**Comment by moonlitSpider** [comment]
Hello again,

I downloaded & installed 1.3.6; set 

      org.osgi.framework.bootdelegation=com.sun.btrace,com.sun.btrace.*,\

per your instruction, but still getting the same result.

How best to debug this?

thanks for your help.

-Paul

**Comment by AbhishekPuligundla** [comment]
Hi, 
     You said you were using java jdk 1.7. Is your application is using this jdk or does your application have its own java. If it has its own jre separately then try to add the tools.jar file or jdk to the application. I explained it above. If you try to copy only tools.jar file make sure to copy it in a bin folder and place that folder in the parent folder of the jre of the application. I hope it works. I think enhancement was already made on this issue. Just check the above comments again.

**Comment by moonlitSpider** [comment]
Thanks, Abishek. I appreciate your help.

I am causing both the Virgo (JVM) application and btrace to use the same Java. Although, to be quite accurate, it's possible that the Virgo app is using the JRE whereas btrace is using the JDK. But in each case it's the same 1.7.0_80 version of Java.

Does this remove the need for the behavior you describe above re tools.jar? 

Another question re the classes named by org.osgi.framework.bootdelegation: do these classes need to be somehow (classpath?) visible to Virgo?

Thanks again.

-Paul

**Comment by moonlitSpider** [comment]
FWIW, here is the -version output. This Java is being used by both btrace & the application that I am trying to trace:

[root@ecxprimary /]# /usr/java/jdk1.7.0_80/bin/java -version
java version "1.7.0_80"
Java(TM) SE Runtime Environment (build 1.7.0_80-b15)
Java HotSpot(TM) 64-Bit Server VM (build 24.80-b11, mixed mode)

**Comment by AbhishekPuligundla** [comment]
As you specified, virgo app is using jre but not the jdk. Try adding the tools.jar file to the virgo jre parent folder(make sure the tools.jar is from same version of java used). I don't remember exactly but btrace uses a particular file from tools.jar file of the target application. The btrace is using the jdk for compiling the btrace script. Why don't try adding tools.jar. I also faced the same kind of problem but after adding tools.jar it started tracing. When we trace a smple java prigram, the btrace and the application running will both have jdk and also jre Environments. There the btrace will consider the tools.jar file from jdk foe the application.

**Comment by AbhishekPuligundla** [comment]
Almost all the applications will be provided with a jre but not jdk since all the files will be .class files so there is no need for jdk. But we are trying to inject byte code to class file usinf the agent concept (brrace is using). So i am not sure but that is one of the reason i looked into adding the file from bin folder of jdk to to target application jre path.

**Comment by moonlitSpider** [comment]
Thanks, Abishek. But I'm not sure how you conclude that Virgo is using the JRE but not the JDK version of the Java executable. In fact, the java executable under ../bin/java is identical to that under ../jre/bin/java (I did a diff).

I should probably make clear that the ../jre directory itself is under /usr/java/**jdk**1.7.0_80/bin. Also there is only one tools.jar and it's at /usr/java/jdk1.7.0_80/lib/tools.jar.

Do you see anything "wrong" with this?

I am more concerned about the meaning and implementation of the org.osgi.framework.bootdelegation=com.sun.btrace,com.sun.btrace.*,\ statement that @jbachorik  mentioned above. Specifically, does Virgo itself need access to the btrace classes....?

-Paul

**Comment by moonlitSpider** [comment]
Here's what "-v" generates

DEBUG: assuming default port 2020
DEBUG: assuming default classpath '.'
DEBUG: reading ./PbTrace.class
DEBUG: read ./PbTrace.class
DEBUG: attaching to 9377

terminating with error:

unable to open socket file: target process not responding or hotspot vm not loaded

**Comment by moonlitSpider** [comment]
One more thing, I am able to monitor this Virgo application using plain old JVisualVM.

Does this fact illuminate things at all?

Thanks for your help.

-Paul

**Comment by jbachorik** [comment]
On May 6, 2016 2:07 PM, "moonlitSpider" <notifications@github.com> wrote:
>
> One more thing, I am able to monitor this Virgo application using plain
old JVisualVM.
>
> Does this fact illuminate things at all?
>
> Thanks for your help.
>
> -Paul
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
<https://github.com/btraceio/btrace/issues/169#issuecomment-217423401>

**Comment by moonlitSpider** [comment]
Is it generally the case JVisualVM for, say, Java 1.X cannot monitor an application compiled by Java 1.Z?

**Comment by jbachorik** [comment]
No. The version of the compiled class does not play any role. The only requirement is that a tools.jar compatible with the target application. This is usually true when you run your application on full JDK instead of JRE (eg. using $JAVA_HOME/bin/java instead of $JAVA_HOME/jre/bin/java)

**Comment by moonlitSpider** [comment]
Very good. 

So, if I run JVisualVM on Windows 7 with my target application on Ubuntu, then I need to ensure that JVisualVM and (by extension, I suppose) btrace are using on Windows a tools.jar file that is compatible (the same as?) the tools.jar on the Ubuntu node?

With respect to the Dockerized OSGI app I am trying to probe, I should point out that a few years ago I was able to use JVisualVM against it **before** it was a Dockerized app. At that time I never tried btrace.

I remain curious as to the meaning of the osgi parameter you mentioned above. Specifically, what is it supposed to do and how does it work? As noted, I did set that bootdelegation parameter but it seems to have had no effect. But that could be because Virgo actually needs access to the classes named in the bootdelegation parameter. But I've no idea.

Thanks again.

Cordially,

Paul
